File: 프로그래머스/2/64065. 튜플/튜플.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <iostream>
#include <algorithm>

using namespace std;

vector<int> solution(string s) {
    vector<int> answer;
    vector<vector<int>> nums;
    
    // 집합 추출
    vector<int> set;
    int num = 0;
    for(int i = 1; i < s.size()-1; i++){
        if(s[i] == '}'){
            set.push_back(num); num = 0; 
            nums.push_back(set); set.clear(); 
        }else if (s[i] == ',' && num != 0){
            set.push_back(num); num = 0; 
        }else if('0' <= s[i] && s[i]  <= '9'){ // 수라면
            num = num*10 + s[i] - '0';
        } 
    }
    
    // 크기 순으로 정렬
    sort(nums.begin(), nums.end(), [](auto &a, auto &b){
         return a.size() < b.size();
    });
    
    // 차례대로 만들기
    unordered_set<int> uni;
    for(int i = 0; i < nums.size(); i++){
        for(int& n : nums[i]){
            if(!uni.count(n)) {
                answer.push_back(n); uni.insert(n);
            }
        }
    }
    
    return answer; // s가 포함하는 튜플을 배열에 담기
}

File: 프로그래머스/2/64065. 튜플/튜플.java
import java.util.*; 

class Solution {
    public int[] solution(String s) {
        
        // 길이 오름차순으로 정렬
        List<List<Integer>> list = new ArrayList<>();
        List<Integer> l = new ArrayList<>();
        int num = 0; 
        for(int i = 1; i < s.length() - 1; i++){
            char cur = s.charAt(i);
            if(cur == '{'){
                l.clear(); 
            }else if(cur == '}'){
                l.add(num); num = 0;
                list.add(new ArrayList<>(l));
                i+=1;
            }else if(Character.isDigit(cur)){
                num = num * 10 + Integer.parseInt(String.valueOf(cur)); 
            }else{ // ,
                l.add(num); num = 0;
                
            }
        }
        
        // 배열 길이로 오름차순 정렬
        list.sort((a, b) -> {
            return Integer.compare(a.size(), b.size()); 
        });
        
        // 정답 배열 만들기
        HashSet<Integer> set = new HashSet<>(); 
        List<Integer> answer = new ArrayList<>();
        for(List<Integer> ll : list){
            for(int n : ll){
                if(!set.contains(n)){
                    answer.add(n); 
                    set.add(n); 
                    break;
                }
            }
        }
        
        
        return answer.stream().mapToInt(i->i).toArray();
    }
}